lgc/builder/InOutBuilder.cpp

```diff
@@ -1257,10 +1258,29 @@ Value *InOutBuilder::readCsBuiltIn(BuiltInKind builtIn, const Twine &instName) {
   }
 
   case BuiltInSubgroupId: {
-    // gl_SubgroupID = gl_LocationInvocationIndex / gl_SubgroupSize
-    Value *localInvocationIndex = readCsBuiltIn(BuiltInLocalInvocationIndex);
-    unsigned subgroupSize = getPipelineState()->getShaderSubgroupSize(m_shaderStage);
-    return CreateLShr(localInvocationIndex, getInt32(Log2_32(subgroupSize)));
+    bool pipelineThreadIdSwizzling =
+        m_pipelineState->getOptions().forceCsThreadIdSwizzling && !getPipelineState()->isComputeLibrary();
+    if (pipelineThreadIdSwizzling) {
+      // From Navi21, it should load the subgroupid from sgpr initialized at wave launch.
```

Comment:
Please just remove this distinction and always use the new-style code. It is better to keep code as orthogonal as possible and keep the implementation of a feature as local as possible. In this case, the implementation of the `forceCsThreadIdSwizzling` toggle can be localized in the implementation of the LocalInvocationId built-in; the toggle doesn't *have* to be taken into account here, and so it shouldn't.

Aside from keeping the impacts of toggles localized, this has two additional benefits:

* There is less code overall.
* An improvement on Navi21/gfx10.3+ for *all* code that uses SubgroupId.

Reply:
Greate Suggestions!! I would like follow it completely.